golioth/golioth.py

```diff
@@ -465,6 +466,69 @@ async def call_method(*args, **kwargs):
 
         return call_method
 
+class DeviceSettings(ApiNodeMixin):
+    def __init__(self, device: Device):
+        self.device = device
+        self.base_url: str = device.base_url
+
+    @property
+    def headers(self) -> Dict[str, str]:
+        return self.device.headers
+
+    async def get_all(self) -> list:
+        response = await self.device.get('settings')
+        return response.json()['list']
+
+    async def get(self, key: str):
+        settings = await self.get_all()
+        for setting in settings:
+            if setting['key'] == key:
+                if 'deviceId' in setting:
+                    return setting
+
+        raise KeyError(f"No setting with {key=}")
+
+    async def set(self, key: str, value: Union[int, float, bool, str],
+                  override: bool = True):
+        if isinstance(value, int):
+            data_type = 'integer'
+        elif isinstance(value, float):
+            data_type = 'float'
+        elif isinstance(value, bool):
+            data_type = 'boolean'
+        elif isinstance(value, str):
+            data_type = 'string'
+        else:
+            raise RuntimeError("Invalid value type")
+
+        json = {
+            "key": key,
+            "dataType": data_type,
+            "value": value,
+            "deviceId": self.device.id,
+        }
+
+        if override:
+            try:
+                setting = await self.device.settings.get(key)
+                response = await self.device.project.put('settings/' + setting['id'], json=json)
+
+                return response.json()
+            except KeyError:
+                pass
```

Comment:
Will this work when `override=True` and there was no such setting previously in backend?

Reply:
It will fall down to line 520 and attempt to POST the setting. This is the same behavior as for project level settings. If the project-level setting exists, it will create a new device level setting. If the project-level setting doesn't exist, then it will get a 500 response from the backend and raise that. I can modify this to handle the 500 better.

Reply:
Added line 526, which makes it so we'll throw a `KeyError` instead of hitting the 500 Response if we attempt to set a device-level setting which doesn't exist in the project.